src/rust/engine/process_execution/src/bounded.rs

```diff
@@ -161,32 +200,47 @@ impl AsyncSemaphore {
   ///
   /// Runs the given Future-creating function (and the Future it returns) under the semaphore.
   ///
+  /// NB: This method does not support preemption, or controlling concurrency.
+  ///
   // TODO: https://github.com/rust-lang/rust/issues/46379
   #[allow(dead_code)]
   pub(crate) async fn with_acquired<F, B, O>(self, f: F) -> O
   where
     F: FnOnce(usize) -> B,
     B: Future<Output = O>,
   {
-    let permit = self.acquire().await;
+    let permit = self.acquire(1).await;
     let res = f(permit.task.id).await;
     drop(permit);
     res
   }
 
-  pub async fn acquire(&self) -> Permit<'_> {
+  ///
+  /// Acquire a slot on the semaphore when it becomes available. Additionally, attempt to acquire
+  /// the given amount of concurrency. The amount actually acquired will be reported on the
+  /// returned Permit.
+  ///
+  pub async fn acquire(&self, concurrency_desired: usize) -> Permit<'_> {
     let permit = self.sema.acquire().await.expect("semaphore closed");
     let task = {
       let mut state = self.state.lock();
       let id = state
         .available_ids
         .pop_front()
         .expect("More permits were distributed than ids exist.");
-      // TODO: Configure and calculate concurrency.
+
+      // A Task is initially given its fair share of the available concurrency (i.e., for two
```

Comment:
I'm not sure the "for two tasks" piece here is accurate - I think the `state.total_permits / (state.tasks.len() + 1)` logic here means that for two tasks each trying to max out concurrency one would initially get the full concurrency, and the second would get half (and the third a third, and so on) until the rebalancing happens? Which I think is necessary, because we can't predict how many tasks _will_ come in, but feels different from "for two tasks, each gets half"?

Reply:
> that for two tasks each trying to max out concurrency one would initially get the full concurrency, and the second would get half (and the third a third, and so on) until the rebalancing happens?

Yea, exactly. The example given was supposed to just indicate that the second task would take half, not that the first task would also get half. I can try to clarify.